packages/core/src/lib/ConnectionManager.ts

```diff
@@ -0,0 +1,336 @@
+   * This is only used when other discovery mechanisms haven't yeild peers that are dialable
+   * This increases relative centralisation of the network
+   */
+  dialAttemptsBeforeBootstrapConnection?: number;
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+/**
+ * ConnectionManager is a singleton class that manages different steps in a libp2p connection lifecycle
+ */
+export class ConnectionManager {
+  private static instance: ConnectionManager;
+  private libp2p: Libp2p;
+  private relay?: IRelay;
+  private pingKeepAliveTimers: {
+    [peer: string]: ReturnType<typeof setInterval>;
+  };
+  private relayKeepAliveTimers: {
+    [peer: string]: ReturnType<typeof setInterval>;
+  };
+  private dialAttempt: number;
+
+  public isConnectionServiceStarted = false;
+
+  public static create(libp2p: Libp2p, options: Options): ConnectionManager {
+    if (!ConnectionManager.instance) {
+      ConnectionManager.instance = new ConnectionManager(libp2p, options);
+    }
+    return ConnectionManager.instance;
+  }
+
+  private constructor(libp2p: Libp2p, options: Options, relay?: IRelay) {
+    this.libp2p = libp2p;
+    this.pingKeepAliveTimers = {};
+    this.relayKeepAliveTimers = {};
+    this.relay = relay;
+    this.dialAttempt = 1;
+
+    this.startDiscoveryConnectionService(
+      options.maxBootstrapPeersAllowed,
+      options.peerDiscoveryIntervalFactor
+    );
+    this.attachEventListeners(options.pingKeepAlive, options.relayKeepAlive);
+  }
+
+  public startDiscoveryConnectionService(
+    maxBootstrapPeersAllowed = DEFAULT_MAX_BOOTSTRAP_PEERS_ALLOWED,
+    interval = DEFAULT_PEER_DISCOVERY_CONNECTION_INTERVAL
+  ): void {
+    if (this.isConnectionServiceStarted) {
+      throw new Error("Connection service already started");
+    }
+
+    this.isConnectionServiceStarted = true;
+
+    this.dialPeersInInterval(maxBootstrapPeersAllowed, interval);
+  }
+
+  public stopAllKeepAlives(): void {
+    for (const timer of [
+      ...Object.values(this.pingKeepAliveTimers),
+      ...Object.values(this.relayKeepAliveTimers),
+    ]) {
+      clearInterval(timer);
+    }
+
+    this.pingKeepAliveTimers = {};
+    this.relayKeepAliveTimers = {};
+  }
+
+  private attachEventListeners(
+    pingKeepAlive: number,
+    relayKeepAlive: number
+  ): void {
+    this.libp2p.connectionManager.addEventListener("peer:connect", (evt) => {
+      this.startKeepAlive(evt.detail.remotePeer, pingKeepAlive, relayKeepAlive);
+    });
+
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
```

Comment:
I don't trust these issues. They need further investigation to assume it's actually the case. Especially that they were opened on old libp2p versions.